notes-20200211.tex

```diff
@@ -0,0 +1,171 @@
+		\infer[no rule]1{\oftype{z{:}A}{M}{C[x \rightarrow z, y \rightarrow z, q \rightarrow \idrefl{z}]}}
+		\infer1[$\idtype$-elim]{\oftype{\idJ[x][y][q][C]{z}{M}{p}}{C[x \rightarrow N, y \rightarrow O, q \rightarrow p]}}
+	\end{prooftree*}
+\item Computation:
+	\begin{prooftree*}
+		\hypo{\istype{A}}
+		\hypo{\oftype{N}{A}}
+		\hypo{\oftype{O}{A}}
+		% FIXME(remexre): same hack as above... :(
+		\infer[no rule]3{\istype{x{:}A,y{:}A,q{:}\idtype{A}{x}{y}}{C}}
+		\infer[no rule]1{\oftype{z{:}A}{M}{C[x \rightarrow z, y \rightarrow z, q \rightarrow \idrefl{z}]}}
+		\infer1[$\idtype$-comp]{\eqterm{\idJ[x][y][q][C]{z}{M}{\idrefl{N}}}{M[z \rightarrow N]}{C[x \rightarrow N, y \rightarrow O, q \rightarrow \idrefl{N}]}}
+	\end{prooftree*}
+\end{enumerate}
+
+The elimination and computation rules here are more complicated than those encountered previously. so some explanation is warranted. These rules state that to prove a statement $C$ containing a term on an identification type $q$, it suffices to consider the case where the term is $\idrefl$. This is justified by the only introduction rule being the reflexive case of equality.
+
+There is also an alternative formulation of $\idtype$-elim and $\idtype$-comp:
+
+% FIXME(remexre): The hack deepens...
+\begin{prooftree*}
+	\hypo{\istype{A}}
+	\hypo{\oftype{M}{A}}
+	\infer[no rule]2{\oftype{P}{\idtype{A}{M}{M}}}
+	\hypo{\istype{x{:}A,q{:}\idtype{A}{M}{x}}{C}}
+	\infer[no rule]1{\oftype{N}{C[x \rightarrow M, q \rightarrow \idrefl{M}]}}
+	\infer2[$\idtype$-elim$^{M=}$]{\oftype{\mathtt{J}^{M=}\brackets{x.q.C}\parens{N; P}}{C[x \rightarrow M, q \rightarrow \idrefl{M}]}}
+\end{prooftree*}
+\begin{prooftree*}
+	\hypo{\istype{A}}
+	\hypo{\oftype{M}{A}}
+	\infer[no rule]2{\oftype{P}{\idtype{A}{M}{M}}}
+	\hypo{\istype{x{:}A,q{:}\idtype{A}{M}{x}}{C}}
+	\infer[no rule]1{\oftype{N}{C[x \rightarrow M, q \rightarrow \idrefl{M}]}}
+	\infer2[$\idtype$-comp$^{M=}$]{\eqterm{\mathtt{J}^{M=}\brackets{x.q.C}\parens{N; \idrefl{M}}}{N}{C[x \rightarrow M, q \rightarrow \idrefl{M}]}}
+\end{prooftree*}
+
+This formulation is often easier to use in practice, but is equivalent to the one initially presented; see section 1.12.1 of \cite{hott-as:book} for details.
+
+\subsection{(Usual) Equality Types}
+
+Next, we introduce equality types $\mathtt{Eq}_A\parens{M; N}$. They are intended to be a direct internalization of judgmental equality by the following equality reflection rule:
+
+\begin{prooftree*}
+	\hypo{\oftype{\ctx}{p}{\mathtt{Eq}_A\parens{M; N}}}
+	\infer1{\eqterm{\ctx}{M}{N}{A}}
+\end{prooftree*}
+
+Note that by the substitutional nature of judgmental equality the converse holds as well:
+
+\begin{prooftree*}
+	\hypo{\eqterm{\ctx}{M}{N}{A}}
+	\infer1{\oftype{\ctx}{\idrefl{M}}{\mathtt{Eq}_A\parens{M; N}}}
+\end{prooftree*}
+
+In this sense, equality types and judgmental equality are equivalent. Formally, the rules of equality types consist of those of identification types and the equality reflection rule.
```

Comment:
I did not (and will not) cover equality types much, but one implementation of equality types additionally have the following rule for _uniqueness:_

If p : Eq_A(M,N) then p=refl : Eq_A(M,N).